File: packages/entity-extraction/android/src/main/java/io/capawesome/capacitorjs/plugins/mlkit/entityextraction/EntityExtractionPlugin.java
package io.capawesome.capacitorjs.plugins.mlkit.entityextraction;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.getcapacitor.Logger;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import io.capawesome.capacitorjs.plugins.mlkit.entityextraction.classes.CustomException;
import io.capawesome.capacitorjs.plugins.mlkit.entityextraction.classes.options.DeleteDownloadedModelOptions;
import io.capawesome.capacitorjs.plugins.mlkit.entityextraction.classes.options.DownloadModelOptions;
import io.capawesome.capacitorjs.plugins.mlkit.entityextraction.classes.options.ExtractEntitiesOptions;
import io.capawesome.capacitorjs.plugins.mlkit.entityextraction.classes.results.ExtractEntitiesResult;
import io.capawesome.capacitorjs.plugins.mlkit.entityextraction.classes.results.GetDownloadedModelsResult;
import io.capawesome.capacitorjs.plugins.mlkit.entityextraction.interfaces.EmptyCallback;
import io.capawesome.capacitorjs.plugins.mlkit.entityextraction.interfaces.NonEmptyResultCallback;
import io.capawesome.capacitorjs.plugins.mlkit.entityextraction.interfaces.Result;

@CapacitorPlugin(name = "EntityExtraction")
public class EntityExtractionPlugin extends Plugin {

    public static final String TAG = "EntityExtraction";

    public static final String ERROR_UNKNOWN_ERROR = "An unknown error occurred.";

    private EntityExtraction implementation;

    @Override
    public void load() {
        implementation = new EntityExtraction();
    }

    @Override
    protected void handleOnDestroy() {
        if (implementation != null) {
            implementation.close();
        }
    }

    @PluginMethod
    public void deleteDownloadedModel(PluginCall call) {
        try {
            DeleteDownloadedModelOptions options = new DeleteDownloadedModelOptions(call);
            EmptyCallback callback = new EmptyCallback() {
                @Override
                public void success() {
                    resolveCall(call);
                }

                @Override
                public void error(@NonNull Exception exception) {
                    rejectCall(call, exception);
                }
            };

            implementation.deleteDownloadedModel(options, callback);
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void downloadModel(PluginCall call) {
        try {
            DownloadModelOptions options = new DownloadModelOptions(call);
            EmptyCallback callback = new EmptyCallback() {
                @Override
                public void success() {
                    resolveCall(call);
                }

                @Override
                public void error(@NonNull Exception exception) {
                    rejectCall(call, exception);
                }
            };

            implementation.downloadModel(options, callback);
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void extractEntities(PluginCall call) {
        try {
            ExtractEntitiesOptions options = new ExtractEntitiesOptions(call);
            NonEmptyResultCallback<ExtractEntitiesResult> callback = new NonEmptyResultCallback<>() {
                @Override
                public void success(@NonNull ExtractEntitiesResult result) {
                    resolveCall(call, result);
                }

                @Override
                public void error(@NonNull Exception exception) {
                    rejectCall(call, exception);
                }
            };

            implementation.extractEntities(options, callback);
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void getDownloadedModels(PluginCall call) {
        try {
            NonEmptyResultCallback<GetDownloadedModelsResult> callback = new NonEmptyResultCallback<>() {
                @Override
                public void success(@NonNull GetDownloadedModelsResult result) {
                    resolveCall(call, result);
                }

                @Override
                public void error(@NonNull Exception exception) {
                    rejectCall(call, exception);
                }
            };

            implementation.getDownloadedModels(callback);
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    private void rejectCall(@NonNull PluginCall call, @NonNull Exception exception) {
        String message = exception.getMessage();
        if (message == null) {
            message = ERROR_UNKNOWN_ERROR;
        }
        String code = null;
        if (exception instanceof CustomException) {
            code = ((CustomException) exception).getCode();
        }
        Logger.error(TAG, message, exception);
        call.reject(message, code);
    }

    private void resolveCall(@NonNull PluginCall call) {
        call.resolve();
    }

    private void resolveCall(@NonNull PluginCall call, @Nullable Result result) {
        if (result == null) {
            call.resolve();
        } else {
            call.resolve(result.toJSObject());
        }
    }
}

File: packages/entity-extraction/ios/Plugin/EntityExtractionPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(EntityExtractionPlugin)
public class EntityExtractionPlugin: CAPPlugin {
    public static let tag = "EntityExtraction"

    private var implementation: EntityExtraction?

    override public func load() {
        implementation = EntityExtraction()
    }

    @objc func deleteDownloadedModel(_ call: CAPPluginCall) {
        do {
            let options = try DeleteDownloadedModelOptions(call)

            implementation?.deleteDownloadedModel(options, completion: { error in
                if let error = error {
                    self.rejectCall(call, error)
                } else {
                    self.resolveCall(call)
                }
            })
        } catch {
            self.rejectCall(call, error)
        }
    }

    @objc func downloadModel(_ call: CAPPluginCall) {
        do {
            let options = try DownloadModelOptions(call)

            implementation?.downloadModel(options, completion: { error in
                if let error = error {
                    self.rejectCall(call, error)
                } else {
                    self.resolveCall(call)
                }
            })
        } catch {
            self.rejectCall(call, error)
        }
    }

    @objc func extractEntities(_ call: CAPPluginCall) {
        do {
            let options = try ExtractEntitiesOptions(call)

            implementation?.extractEntities(options, completion: { result, error in
                if let error = error {
                    self.rejectCall(call, error)
                } else {
                    self.resolveCall(call, result)
                }
            })
        } catch {
            self.rejectCall(call, error)
        }
    }

    @objc func getDownloadedModels(_ call: CAPPluginCall) {
        let modelIdentifiers = implementation?.getDownloadedModels() ?? []
        let result = GetDownloadedModelsResult(modelIdentifiers: modelIdentifiers)
        self.resolveCall(call, result)
    }

    private func rejectCall(_ call: CAPPluginCall, _ error: Error) {
        CAPLog.print("[", EntityExtractionPlugin.tag, "] ", error)
        let code = (error as? CustomError)?.code
        call.reject(error.localizedDescription, code)
    }

    private func resolveCall(_ call: CAPPluginCall) {
        call.resolve()
    }

    private func resolveCall(_ call: CAPPluginCall, _ result: Result?) {
        if let result = result?.toJSObject() as? JSObject {
            call.resolve(result)
        } else {
            call.resolve()
        }
    }
}
